docs/en/sphinx-doc/11_Appendix/01_Configuration.rst

設定ファイルの定義
=============

**cre_config.pl** コマンドにて$GETPERF_HOME/confの下に各設定ファイルが生成されます。
これら設定はインストール作業時に

getperf_site.json
-----------------

Getperf のベース設定と、各種インストールソフトウェアのプロパティ設定を行います。Getperf ホームディレクトリ、ログ出力設定をします。

+-------------------------------+-------------------------------+-------------------------------------------------------------------------------------------+
|項目名                         |既定値                         |定義                                                                                       |
+===============================+===============================+===========================================================================================+
|GETPERF_HOME                   |COMPONENT_ROOT                 |Getperfのホームディレクトリ。デフォルトは管理ユーザのホームディレクトリの下のgetperf       |
+-------------------------------+-------------------------------+-------------------------------------------------------------------------------------------+
|GETPERF_SITE_DIR               |COMPONENT_ROOT/t               |ユニットテスト用サイトディレクトリ                                                         |
+-------------------------------+-------------------------------+-------------------------------------------------------------------------------------------+
|GETPERF_STAGING_DIR            |COMPONENT_ROOT/t/staging_data  |ユニットテスト用グラフデータディレクトリ                                                   |
+-------------------------------+-------------------------------+-------------------------------------------------------------------------------------------+
|GETPERF_TMPFS_DIR              |/tmp                           |RRDtoolのデータキャッシュディレクトリ                                                      |
+-------------------------------+-------------------------------+-------------------------------------------------------------------------------------------+
|GETPERF_LOG_DIR                |COMPONENT_ROOT/log             |ログ出力ディレクトリ                                                                       |
+-------------------------------+-------------------------------+-------------------------------------------------------------------------------------------+
|GETPERF_LOG_LEVEL              |notice                         |デーモンプロセスのログレベル。debug,info,notice,warn,err,crit から選択                     |
+-------------------------------+-------------------------------+-------------------------------------------------------------------------------------------+
|GETPERF_STDOUT_LOG_LEVEL       |notice                         |コマンド実行のログレベル。debug,info,notice,warn,err,crit から選択                         |
+-------------------------------+-------------------------------+-------------------------------------------------------------------------------------------+
|GETPERF_DISK_UTIL_LIMIT_MB     |100                            |ディスク容量アラームの閾値(MB)。ディスク空き容量が本値を下回ると致命的エラーで終了します   |
+-------------------------------+-------------------------------+-------------------------------------------------------------------------------------------+

Redis サーバのキュー設定を行います。クラスター構成のノード間の通信に使用します。

+-----------------------+-----------------------+-----------------------+
|項目名                 |既定値                 |定義                   |
+=======================+=======================+=======================+
|GETPERF_REDIS_PORT     |6379                   |Redis サービスポート   |
+-----------------------+-----------------------+-----------------------+
|GETPERF_REDIS_HOST     |localhost              |Redis サービスホスト   |
+-----------------------+-----------------------+-----------------------+
|GETPERF_QUEUE          |getperf:queue:default  |Redis キュー名         |
+-----------------------+-----------------------+-----------------------+

MySQL のパスワード、SSL証明書の設定をします。

+-----------------------------------+-------------------+-----------------------------------+
|項目名                             |既定値             |定義                               |
+===================================+===================+===================================+
|GETPERF_CACTI_MYSQL_ROOT_PASSWD    |getperf            |MySQLのルートパスワード            |
+-----------------------------------+-------------------+-----------------------------------+
|GETPERF_ADMIN_SSL_DIR              |/etc/getperf/ssl   |SSL証明書ディレクトリ              |
+-----------------------------------+-------------------+-----------------------------------+
|GETPERF_SSL_EXPIRATION_DAY         |365                |SSL証明書有効期限。日数を指定      |
+-----------------------------------+-------------------+-----------------------------------+
|GETPERF_LICENSE_POLICY             |none               |ライセンスポリシー(現在はnoneのみ) |
+-----------------------------------+-------------------+-----------------------------------+

エージェント Webサービスの設定で、Apache HTTP サーバと、Tomcat Webコンテナの設定をします。Apache, Tomcat の Web　サービスは 管理用の admin と、データ通信用の data の 2つで構成します。初めに管理用 Web サービス設定をします。

+-------------------------------+-------------------------------+-----------------------------------------------+
|項目名                         |既定値                         |定義                                           |
+===============================+===============================+===============================================+
|GETPERF_WS_TOMCAT_OWNER        |${admin_user}                  |エージェントWebサービスプロセスオーナー        |
+-------------------------------+-------------------------------+-----------------------------------------------+
|GETPERF_WS_SERVER_NAME         |${local_ip}                    |廃止。エージェントWebサービスホスト            |
+-------------------------------+-------------------------------+-----------------------------------------------+
|GETPERF_WS_SERVER_ADMIN        |you@example.com                |エージェントWebサービス管理者連絡先            |
+-------------------------------+-------------------------------+-----------------------------------------------+
|GETPERF_WS_PROTOCOL            |https                          |httpまたはhttps                                |
+-------------------------------+-------------------------------+-----------------------------------------------+
|GETPERF_WS_ADMIN_SUFFIX        |admin                          |インストールディレクトリのサフィックス。       |
|                               |                               |tomcat-{suffix}となる                          |
+-------------------------------+-------------------------------+-----------------------------------------------+
|GETPERF_WS_ADMIN_SERVER        |${local_ip}                    |サーバ名                                       |
+-------------------------------+-------------------------------+-----------------------------------------------+
|GETPERF_WS_ADMIN_PORT_SSL      |57443                          |Apache HTTPSポート                             |
+-------------------------------+-------------------------------+-----------------------------------------------+
|GETPERF_WS_ADMIN_PORT_RUN      |57005                          |Tomcat 管理ポート                              |
+-------------------------------+-------------------------------+-----------------------------------------------+
|GETPERF_WS_ADMIN_PORT_AJP      |57009                          |Apache AJPポート                               |
+-------------------------------+-------------------------------+-----------------------------------------------+
|GETPERF_WS_ADMIN_PORT          |57000                          |Apache HTTPポート                              |
+-------------------------------+-------------------------------+-----------------------------------------------+
|GETPERF_WS_ADMIN_DIR           |/axis2/services/GetperfService |エージェントWebサービス URL                    |
+-------------------------------+-------------------------------+-----------------------------------------------+

データ通信用 Apache, Tomcat Web サービス設定をします。

+---------------------------+-------------------------------+-----------------------------------------------+
|項目名                      |既定値                        |定義                                           |
+===========================+===============================+===============================================+
|GETPERF_WS_DATA_SUFFIX     |data                           |インストールディレクトリのサフィックス。       |
|                           |                               |tomcat-{suffix}となる                          |
+---------------------------+-------------------------------+-----------------------------------------------+
|GETPERF_WS_DATA_SERVER     |${local_ip}                    |Webサービスホスト                              |
+---------------------------+-------------------------------+-----------------------------------------------+
|GETPERF_WS_DATA_PORT_SSL   |58443                          |Apache HTTPS ポート                            |
+---------------------------+-------------------------------+-----------------------------------------------+
|GETPERF_WS_DATA_PORT_RUN   |58005                          |Tomcat 管理ポート                              |
+---------------------------+-------------------------------+-----------------------------------------------+
|GETPERF_WS_DATA_PORT_AJP   |58009                          |Apache AJP ポート                              |
+---------------------------+-------------------------------+-----------------------------------------------+
|GETPERF_WS_DATA_PORT       |58000                          |Apache HTTPポート                              |
+---------------------------+-------------------------------+-----------------------------------------------+
|GETPERF_WS_DATA_DIR        |/axis2/services/GetperfService |インストールディレクトリサフィックス           |
+---------------------------+-------------------------------+-----------------------------------------------+

Apache, Tomcat Webサービスの共通設定をします。

+---------------------------+-------------------+------------------------------------+
|項目名                     |既定値             |定義                                |
+===========================+===================+====================================+
|GETPERF_WS_APACHE_HOME     |/usr/local/apache  | Apache インストールディレクトリ    |
+---------------------------+-------------------+------------------------------------+
|GETPERF_WS_TOMCAT_HOME     |/usr/local/tomcat  | Tomcat インストールディレクトリ    |
+---------------------------+-------------------+------------------------------------+
|GETPERF_WS_MAX_SERVERS     |30                 | Tomcat 最大スレッド数(現在未使用)  |
+---------------------------+-------------------+------------------------------------+
|GETPERF_WS_TIMEOUT         |30                 | Tomcat タイムアウト秒(現在未使用)  |
+---------------------------+-------------------+------------------------------------+
|GETPERF_WS_WAIT_TIME       |60                 | 現在未使用                         |
+---------------------------+-------------------+------------------------------------+

エージェントデータ転送で転送されたzipファイルの設定をします。

+-------------------------------+-------+-------------------------------------------+
|項目名                         |既定値 |定義                                       |
+===============================+=======+===========================================+
|GETPERF_ZIP_PURGE_ENABLE       |TRUE   | 過去のzipファイルの消込処理を有効にするか |
+-------------------------------+-------+-------------------------------------------+
|GETPERF_ZIP_RETENTION_HOUR     |1      | 過去のzipファイルの保存期間(時間)         |
+-------------------------------+-------+-------------------------------------------+
|GETPERF_ZIP_PURGE_HOUR         |3      | 過去のzipファイルの削除期間(時間)         |
+-------------------------------+-------+-------------------------------------------+

エージェントモジュールの配置、バージョンの設定をします。

+-------------------------------+---------------------------+--------------------------------------------------------------+
|項目名                          |既定値                    |定義                                                          |
+===============================+===========================+==============================================================+
|GETPERF_AGENT_TAR              |COMPONENT_ROOT/var/agent   | エージェントモジュールダウンロードサイトのホームディレクトリ |
+-------------------------------+---------------------------+--------------------------------------------------------------+
|GETPERF_AGENT_MAJOR_VERSION    |2                          | エージェントのメジャーバージョン                             |
+-------------------------------+---------------------------+--------------------------------------------------------------+

集計データの保持期間を設定します。

+----------------------------+------------------+------------------------------------------------------------------------------+
|項目名                       |既定値           |定義                                                                          |
+============================+==================+==============================================================================+
|GETPERF_PURGE_DATA_HOUR     |"analysis": 3     | zip解凍したエージェント受信データの保存ディレクトリ anlysis の保持期間(時間) |
+----------------------------+------------------+------------------------------------------------------------------------------+
|GETPERF_PURGE_DATA_HOUR     |"summary": 3      | 集計データの保存ディレクトリ summary の保持期間(時間)                        |
+----------------------------+------------------+------------------------------------------------------------------------------+

getperf_cacti.json
------------------

グラフモニタリングツール Cacti の配置、バージョンの設定をします。

+---------------------------+-------------------------------+---------------------------------------------------------------------------------------------+
|項目名                      |既定値                            |定義                                                                                     |
+===========================+===============================+=============================================================================================+
|GETPERF_CACTI_HTML         |/var/www/html                  | Cacti ホームのルートディレクトリ。各サイトのCacti ホームは本ディレクトリの下にリンクされる  |
+---------------------------+-------------------------------+---------------------------------------------------------------------------------------------+
|GETPERF_CACTI_ARCHIVE_DIR  |[%COMPONENT_ROOT%]/var/cacti   | Cacti モジュールアーカイブの保存先                                                          |
+---------------------------+-------------------------------+---------------------------------------------------------------------------------------------+
|GETPERF_CACTI_ARCHIVE      |cacti-0.8.8e.tar.gz            | Cacti モジュールアーカイブのファイル名                                                      |
+---------------------------+-------------------------------+---------------------------------------------------------------------------------------------+
|GETPERF_CACTI_HOME         |[%COMPONENT_ROOT%]/lib/cacti   | 各さーびの Cacti ホームディレクトリ                                                         |
+---------------------------+-------------------------------+---------------------------------------------------------------------------------------------+
|GETPERF_CACTI_TEMPLATE_DIR |template/0.8.8e                | Cacti テンプレートの保存先                                                                  |
+---------------------------+-------------------------------+---------------------------------------------------------------------------------------------+
|GETPERF_CACTI_DUMP         |template/0.8.8e/cacti.dmp      | Cacti リポジトリデータベースのダンプファイル                                                |
+---------------------------+-------------------------------+---------------------------------------------------------------------------------------------+
|GETPERF_CACTI_CONFIG       |template/config.php.tpl        | Cacti 設定ファイルのテンプレートの保存先                                                    |
+---------------------------+-------------------------------+---------------------------------------------------------------------------------------------+

getperf_rrd.json
----------------

時系列データベース RRDtool のリテンション、集計期間の設定をします。

+-----------+-------+---------------------------------------------------------------------------+
|項目名     |既定値 |定義                                                                       |
+===========+=======+===========================================================================+
|label      |Dayly  | 直近1日のグラフを参照するときの設定。以下、step,save_daysの設定に従います |
+-----------+-------+---------------------------------------------------------------------------+
|step       |120    | 集計期間(秒)                                                              |
+-----------+-------+---------------------------------------------------------------------------+
|save_days  |1      | データ保持期間(日)                                                        |
+-----------+-------+---------------------------------------------------------------------------+
|label      |Weekly | 直近1週間のグラフを参照するときの設定                                     |
+-----------+-------+---------------------------------------------------------------------------+
|step       |900    | 集計期間(秒)                                                              |
+-----------+-------+---------------------------------------------------------------------------+
|save_days  |8      | データ保持期間(日)                                                        |
+-----------+-------+---------------------------------------------------------------------------+
|label      |Monthly| 直近1カ月のグラフを参照するときの設定                                     |
+-----------+-------+---------------------------------------------------------------------------+
|step       |3600   | 集計期間(秒)                                                              |
+-----------+-------+---------------------------------------------------------------------------+
|save_days  |31     | データ保持期間(日)                                                        |
+-----------+-------+---------------------------------------------------------------------------+
|label      |Yearly | 直近1年のグラフを参照するときの設定                                       |
+-----------+-------+---------------------------------------------------------------------------+
|step       |86400  | 集計期間(秒)                                                              |
+-----------+-------+---------------------------------------------------------------------------+
|save_days  |730    | データ保持期間(日)                                                        |
+-----------+-------+---------------------------------------------------------------------------+

getperf_rsync.json
------------------

別サーバのエージェント転送データをフォワードするときの設定で、特定サイトのデータを別サーバに複製するときに使用します。また、旧パージョン(V1)のサイトを現バージョン(V2)に移行する際に使用します。本設定はオプションとなり、デフォルトは無効となります。設定の詳細については、[旧バージョンのサイトの移行](docs/ja/docs/10_Miscellaneous/02_PotingOldVersion.md)を参照してください

+---------------------------+-------------------+-------------------------------------------+
|項目名                      |既定値            |定義                                       |
+===========================+===================+===========================================+
|GETPERF_RSYNC_SOURCE       |staging_data/site  | RSync ソース名                            |
+---------------------------+-------------------+-------------------------------------------+
|GETPERF_RSYNC_ZIP_KEYWORD  |                   | データ転送ファイルのキーワードフィルター  |
+---------------------------+-------------------+-------------------------------------------+
|GETPERF_RSYNC_HOST         |localhost          | データ転送基のホスト                      |
+---------------------------+-------------------+-------------------------------------------+

getperf_zabbix.json
-------------------

オープンソースの統合監視ソフト Zabbix の設定をします。本設定はオプションで、デフォルトは無効となります。

+---------------------------+---------------------------------------------------------------------------------------------------+-------------------------------------------------------------------+

Changes.txt
Changes for 2.13.0 (2016-8-21)
--------------------------------------------------------------------------------
New features:
* Zabbix 3.0のサポート
* 監視サーバのHA化

Bug fixes:

Changes for 2.12.0 (2016-6-20)
--------------------------------------------------------------------------------
New features:
* Cacti 管理コマンド(cacti-cli)に合計線のオプション追加
* 集計デーモン異常終了時にモニタースクリプト追加
* 各種テンプレートの初版作成。Linux / Windows / JavaVM / TTSP / Oracle
* Zabbix 管理コマンド(zabbix-cli)に Zabbix データ転送 API 追加
* Zabbix 監視ホスト、アイテム定義ファイルのパーサー追加
* Cacti MySQL DB への SQL API 追加
* Cacti 管理コマンド(cacti-cli)に node_alias オプション追加

Bug fixes:
* 集計デーモン起動スクリプトで2つ目のサイト集計で停止する問題対処
* tomcatログローテーションパス名指定誤りの対処
* 監視エージェントの全てのコマンドが周期実行の場合に途中で採取を停止してしまう問題対処

Changes for 2.11.0 (2016-3-21)
--------------------------------------------------------------------------------
New features:
* Cacti 管理コマンド(cacti-cli)にDBバックアップ/リストアオプション追加

Bug fixes:
* Webサービス既定ログレベルをDEBUGからINFOに下げた
* エージェントminizipソースのディレクトリ分割
* サイト初期化時にssh の確認メッセージをスキップするように変更

Changes for 2.10.0 (2016-3-7)
--------------------------------------------------------------------------------
New features:
* 複数監視サーバ構成での Zabbix 設定
* ストレージエンジンとして InfluxDB をサポート(詳細は、 docs/ja/setup/influxdb.md を参照)

Bug fixes:
* Webサービス用ディレクトリ(t/staging_data)をコピーしないで本体を移行すると、zip送信でAtachement Errrorが発生する
* 1K未満の小さい採取ファイルで集計処理が実行されない

Changes for 2.9.0 (2016-3-3)
--------------------------------------------------------------------------------
New features:
* OSSコミュニティ版ソースコード改修
  - GPLv2 ライセンスファイル作成
  - gSOAP ソースの分離
  - ユーザガイドの英語化(機械翻訳)
  - ユニットテスト手順作成
  - 使用OSSのライセンスファイル作成
  - 採取エージェントの未使用コード削除

Changes for 2.8.5 (2016-2-7)
--------------------------------------------------------------------------------
New features:
* RRDtool 管理コマンド(rrd-cli)追加。RRA データの追加、削除。RRD ファイルの複製

Buf fixes:
* グラフ登録コマンドでテナント指定時に既存(_default)の定義が削除される問題の対処
* Zabbix2.4以降でZabbix管理コマンドがログイン認証エラーとなる問題対処

Changes for 2.8.4 (2016-1-25)
--------------------------------------------------------------------------------
New features:
* ビュー定義ファイルのパスを view/{tenant}/{domain}/{node}.json に変更
* Cactiグラフメニューの登録でソート機能追加
* RRDtool キャッシュ機能追加。非同期で RRDtool 書き込みを行う
* 集計結果のタイムスタンプ列を文字列からエポック値(秒)に変更

Buf fixes:
* Linux uname -a コマンド集計スクリプトのエラー対処
* RRDtoolロードモジュールのパスをLoaderの下に移動

Changes for 2.8.3 (2016-1-11)
--------------------------------------------------------------------------------
New features:
* マニュアルを reST フォーマットに変更。Sphinx 導入
* SSL ルート認証局と中間認証局の2段構成に変更

Buf fixes:
* RRDtoolの綴りを修正

Changes for 2.8.2 (2015-11-30)
--------------------------------------------------------------------------------
New features:
* マニュアル初版作成
* Oracle STATSPACKテンプレート作成
* グラフ登録でデバイスのソートオプション追加
* グラフ登録で強制更新オプションにメニュー更新無効化オプションを追加

Buf fixes:
* JavaVMヒープモニタリングでログが標準出力されない問題対処

Changes for 2.8.1 (2015-11-4)
--------------------------------------------------------------------------------
New features:
* 開発サイトのサイト更新手順。git clone してから、initsite.pl {site_dir} --update でサイトを更新する
* ストレージエンジンとして Graphite をサポート(詳細は、 docs/ja/setup/graphite.md を参照)
* vCenter テンプレートで、Linux から ODBC 経由で vCenter 情報を採取する設定を追加

Bug fixes:
* データ集計コマンドで、サイト設定ファイル内のサイトホームが実際のディレクトリと異なる場合に集計がされない問題対処
* log/getperf.log のログローテーション追加
* サイト初期化の更新モードでサイトディレクトリが存在しない場合の例外追加
* root 実行時に getperf.log のオーナがrootになる問題対処

Changes for 2.8.0 (2015-10-7)
--------------------------------------------------------------------------------
New features:
* Solaris プラットフォームのエージェントサポート
* Zabbix GUI のスクリプト化。グループ、ホスト、テンプレートの自動登録を追加
* Rex 連携スクリプト。指定ノードで Rex 操作を行う
* rex make_agent_src でエージェントのビルド番号(n)、リリース番号(2.x.x)の管理ファイルに一元化
	ビルド番号 getperf-agent/BUILD で管理
	リリース番号 getperf/RELEASE で管理
* エージェントソース作成でネットワーク設定ファイル一式の部分ダウンロードファイルを追加

Bug fixes:
* エージェントのディレクトリ作成処理で事前にディレクトリの有無をチェックする
* Solaris 環境エージェントの zip 圧縮でディレクトリ判別に失敗する問題対処
* Windows 環境エージェントのディレクトリチェックに失敗する問題対処
* gccの古いバージョン(2008年頃)で、BOM 付きの UTF-8 ソースコードでコンパイルエラーとなる問題のワークアラウンド手順追記

Changes for 2.7.3 (2015-07-17)
--------------------------------------------------------------------------------

New features:
* vCenter VMWare 仮想化環境収集テンプレート VCenter 追加
* エージェントソースアーカイブの作成スクリプト追加
* Zabbix のインストールスクリプト追加
* エージェントバイナリのダウンロードサイト設定追加

Bug fixes:
* エージェント起動時のデーモンプロセス化で不要なシグナルを受付けない様に変更
* Javaモニタリング時にJVMリスト出力で NullPointerException が発生する問題対応
* Cacti 管理スクリプトで メッセージフラッシュが効かない問題対処
* sumupデーモン起動で、標準出力、エラー出力をサイトホーム下にログ出力するように変更
* インストールのワークディレクトリ /tmp/Rex のオーナーが root になる問題対処
* Cacti 管理スクリプトのグラフテンプレート登録でアイテムが二重登録される問題対処
* Cacti 管理スクリプトでデバイス名に特殊文字が含まれる場合にSQL実行でエラーとなる問題対処

Changes for 2.7.0 (2015-07-20)
--------------------------------------------------------------------------------

New features:
* Cacti GUI のスクリプト化。グラフテンプレート自動作成機能追加
* Cacti-0.8.8e にアップグレード
* ストレージ収集テンプレート Storage 追加
* JavaVM 収集テンプレート JVM 追加

Bug fixes:
* 数百MB以上の採取ファイルのzip送信でヒープ不足エラー発生
* JavaVM収集モジュールで、対象VMの実行コマンドがnullの場合に解析エラーになる問題対処
* インストール後、初回の情報採取で、staging_data に jsonリストがないために、スキップされる問題対処
* インストールスクリプトの更新。libxml2-devel の yum インストール追加。axis2 モジュールの URL 変更
* グラフ登録コマンドで出力メッセージがフラッシュされない問題対処
* テンプレート化により不要となったsitelibディレクトリ削除
* サイトコマンドのリンクの廃止。合わせてSiteInfo::switch_site_command_lib_link 削除

Changes for 2.6.3 (2015-06-25)
--------------------------------------------------------------------------------

New features:
* 旧サイトとの同期機能 sitesync 追加(旧サイトで rsyncd 設定が必要)。
  sitesync rsync://xxx/yyy を定期実行することで、旧サイトのzipファイルを転送、解凍して集計する
* Linux Load average, Process 集計スクリプト追加。
* データ集計に標準出力関数 standard_report() 追加。
* Cacti クライアントツールにホストテンプレートの export/import 機能追加。
* ドメインの export/import 機能追加。

Bug fixes:
* サイト初期化後の初回のzip解凍でsumup startしないと、sumup --recoverが効かない問題対処
* エージェント EUC_JP のヘルプメッセージのタイポミス
* sitesync スクリプト追加により不要となった rsync.pl の廃止
* Linux ネットワーク /proc/net/dev の解析不具合対処。サマリを集計する場合、複数のCOUNTERの合算値計算は問題があるため、代表デバイスを選択する方式に変更。
* Windows パフォーマンスモニタ Total インスタンス値が　0　として誤集計される問題対処
* Linux ディスクI/O iostat 結果のタイムスタンプのご集計の問題対処
* rrdtool update で800行以上の更新でrrdtoolがサスペンドする問題対処。指定行に分割して更新する

Changes for 2.6.2 (2015-05-16)
--------------------------------------------------------------------------------

New features:
* sumup --daemon 追加。サイト毎にデーモンプロセスを立ち上げる。
* sumupctlスクリプト追加。OS起動時に、全サイトのデーモンプロセスを立ち上げる。

lib/Getperf/Config.pm
use strict;
use warnings;
package Getperf::Config;

use Data::Dumper;
use Log::Handler app => "LOG";
use Log::Handler::Output::File::Stamper;
use File::Basename qw(dirname);
use File::Copy;
use Object::Container -base;
use FindBin;
use Path::Class;
use JSON::XS;
use File::Basename qw(dirname);
use constant {
	PATH_TIME_STAMP => 1,
	PATH_TIME_SHIFT => 2,
};

our $GETPERF_HOME     = $ENV{'GETPERF_HOME'} ||file(dirname(__FILE__) . '/../..')->absolute->resolve->stringify;
our $GETPERF_CONF_DIR = $ENV{'GETPERF_CONF_DIR'} || "$GETPERF_HOME/config";
our $AUTO_AGGREGATE   = 1;
our $AUTO_DEPLOY      = 1;
our $GETPERF_INIT;

sub read_config {
	my ($json_file) = @_;
	my $config_file = file($GETPERF_CONF_DIR, $json_file);
	if ($config_file->stat) {
		my $config_json_text = $config_file->slurp || die $@;
    	return decode_json($config_json_text);
	}
	return {};
}

sub read_sitekeys {
	my @sites = ();
	my @site_lists = dir($GETPERF_CONF_DIR, 'site')->children;
	for my $site ( sort @site_lists ) {
		next if ($site->basename!~/^(.+?)\.json/);
    	push(@sites, $1);
	}
	return @sites;
}

sub read_site_config {
	my ($sitekey) = @_;
	return read_config("site/$sitekey.json");
}

sub dump_site_config_json {
	my ($sitekey, $c) = @_;

	sub nvl {
		my ($src, $default) = @_;
		return (defined($src))?$src:$default;
	}

	my @json = (
		'{' ,
		'	"site_key":   "' . $sitekey . '",' ,
		'	"access_key": "' . nvl($c->{access_key}, '') . '",' ,
		'	"home":       "' . nvl($c->{home}, '') . '",' ,
		'	"user":       "' . nvl($c->{user}, '') . '",' ,
		'	"group":      "' . nvl($c->{group}, '') . '",' ,
		'',
		'	"auto_aggregate": ' . nvl($c->{auto_aggregate}, $AUTO_AGGREGATE) . ',' ,
		'	"auto_deploy":    ' . nvl($c->{auto_deploy}, $AUTO_DEPLOY) ,
		'}',
	);

	return join("\n", @json);
}

sub write_site_config {
	my ($sitekey, $c) = @_;

	my $config = "$GETPERF_CONF_DIR/site/${sitekey}";
	my $config_json = $config . '.json';
	my $config_tmp  = $config . '.tmp';
	my $config_bak  = $config . '.bak';

	my $config_file = file($config_tmp);
	my $writer = file($config_file)->open('w') || die "$! : $config_file";
	my $buf = dump_site_config_json($sitekey, $c);
	$writer->print($buf);
	$writer->close;

	if (-f $config_json) {
		move $config_json, $config_bak || die $!;
	}
	move $config_tmp,  $config_json || die $!;

	return 1;
}

{
	package BaseConfig;
	use Log::Handler app => "LOG";
	use Data::Dumper;

	sub new {
	    my ($class) = @_;

		my $config = Getperf::Config::read_config('getperf_site.json');
	    my $self = bless {
		    root_instance      => 1, 
			config_file        => "$GETPERF_CONF_DIR/getperf_site.conf",
			config_dir         => $GETPERF_CONF_DIR,
			site_config_dir    => "$GETPERF_CONF_DIR/site",
			debug              => 0,
			fork               => 1,
			max_processes      => 16,
			sitekey            => undef,
			home               => $config->{GETPERF_HOME} || $GETPERF_HOME,
			ssl_admin_dir      => $config->{GETPERF_ADMIN_SSL_DIR} || "/etc/getperf/ssl",
			ssl_expiration_day => $config->{GETPERF_SSL_EXPIRATION_DAY} || 365,
			ssl_root_ca        => $config->{GETPERF_SSL_COMMON_NAME_ROOT_CA} || 'getperf_ca',
			ssl_inter_ca       => $config->{GETPERF_SSL_COMMON_NAME_INTER_CA} || 'getperf_inter',
			license_policy     => $config->{GETPERF_LICENSE_POLICY} || 'none',
			ws_tomcat_owner    => $config->{GETPERF_WS_TOMCAT_OWNER} || 'pscommon',
			ws_server_name     => $config->{GETPERF_WS_SERVER_NAME} || '127.0.0.1',
			ws_server_admin    => $config->{GETPERF_WS_SERVER_ADMIN} || 'you@example.com',
			ws_protocol        => $config->{GETPERF_WS_PROTOCOL} || 'https',
			ws_apache_dir      => $config->{GETPERF_WS_APACHE_HOME} || "/usr/local/apache",
			ws_tomcat_dir      => $config->{GETPERF_WS_TOMCAT_HOME} || "/usr/local/tomcat",
			ws_admin_server    => $config->{GETPERF_WS_ADMIN_SERVER} || '127.0.0.1',
			ws_admin_suffix    => $config->{GETPERF_WS_ADMIN_SUFFIX} || "admin",
			ws_admin_port_ssl  => $config->{GETPERF_WS_ADMIN_PORT_SSL} || 57443,
			ws_admin_port_run  => $config->{GETPERF_WS_ADMIN_PORT_RUN} || 57005,
			ws_admin_port_ajp  => $config->{GETPERF_WS_ADMIN_PORT_AJP} || 57009,
			ws_admin_port      => $config->{GETPERF_WS_ADMIN_PORT} || 57000,
			ws_admin_dir       => $config->{GETPERF_WS_ADMIN_DIR} || '/axis2/services/GetperfService',
			ws_data_server     => $config->{GETPERF_WS_DATA_SERVER} || '127.0.0.1',
			ws_data_suffix     => $config->{GETPERF_WS_DATA_SUFFIX} || "data",
			ws_data_port_ssl   => $config->{GETPERF_WS_DATA_PORT_SSL} || 58443,
			ws_data_port_run   => $config->{GETPERF_WS_DATA_PORT_RUN} || 58005,
			ws_data_port_ajp   => $config->{GETPERF_WS_DATA_PORT_AJP} || 58009,
			ws_data_port       => $config->{GETPERF_WS_DATA_PORT} || 58000,
			ws_data_dir        => $config->{GETPERF_WS_DATA_DIR} || '/axis2/services/GetperfService',
			staging_dir        => $config->{GETPERF_STAGING_DIR} || "$GETPERF_HOME/staging",
			site_dir           => $config->{GETPERF_SITE_DIR} ||    "$GETPERF_HOME/site",
			tmpfs_dir          => $config->{GETPERF_TMPFS_DIR} ||   "$GETPERF_HOME/tmpfs",
			lib_dir            => $config->{GETPERF_LIB_DIR} ||     "$GETPERF_HOME/lib",
			log_dir            => $config->{GETPERF_LOG_DIR} ||     "$GETPERF_HOME/log",
			agent_tar          => $config->{GETPERF_AGENT_TAR} ||   "$GETPERF_HOME/var/agent",
			agent_major_ver    => $config->{GETPERF_AGENT_MAJOR_VERSIOPN} || 2,
			log_level          => $config->{GETPERF_LOG_LEVEL} ||   'info', 
			stdout_log_level   => $config->{GETPERF_STDOUT_LOG_LEVEL} ||   'error', 
			disk_util_limit    => $config->{GETPERF_DISK_UTIL_LIMIT_MB} || 100,
			purge_data_hour    => $config->{GETPERF_PURGE_DATA_HOUR} || undef,
			admin_schedule     => $config->{GETPERF_ADMIN_SCHEDULE} || undef,
			mysql_passwd       => $config->{GETPERF_CACTI_MYSQL_ROOT_PASSWD} || undef,
	    }, $class;
	    $self->_init_log if (!$GETPERF_INIT);
		$GETPERF_INIT = 1;
	    return $self;
	}

	sub _init_log {
		my ($self) = @_;

		LOG->add(
			'Log::Handler::Output::File::Stamper' => +{
				filename       => $self->{log_dir} . '/getperf.log.%d{yyyyMMdd}',
				maxlevel       => $self->{log_level},
				permissions    => "0664",
				timeformat     => "%Y/%m/%d %H:%M:%S",
				message_layout => "%T [%L] %m",
			},
		);
		# Log owner to change the owner it was written in the config. 
		# problems that owners become root when root run.
		my $owner = $self->{ws_tomcat_owner};
		eval {
			my ($uid, $gid) = (getpwnam $owner)[2,3] or die "getpwnam $owner";
			if (my $filename = LOG->{outputs}[0]->{output}->{filename}) {
				chown $uid, $gid, $filename;
			}
		};
	}

	sub add_screen_log {
		my ($self) = @_;
		LOG->add(
			screen => {
				log_to   => "STDOUT",
				maxlevel => $self->{stdout_log_level},
				timeformat      => "%Y/%m/%d %H:%M:%S",
				message_layout => "%T [%L] %m",
			},
		);
	}
}

{
	package SiteConfig;

	sub new {
	    my ($class) = @_;

		
		my @config = Getperf::Config::read_sitekeys();
	    bless {
	    	sites => \@config,
	    }, $class;
	}
}

{
	package RsyncConfig;

	sub new {
	    my ($class) = @_;

		my $config = Getperf::Config::read_config('getperf_rsync.json') || undef;
	    bless {
	    	($config)? %$config : (),
	    }, $class;		
	}
}

{
	package RRDConfig;

	sub new {
	    my ($class) = @_;

		my $config = Getperf::Config::read_config('getperf_rrd.json');
	    bless {
	    	rra => $config,
	    }, $class;
	}
}

{
	package QueueConfig;

	sub new {
	    my ($class) = @_;

		my $config = Getperf::Config::read_config('getperf_queue.json');
	    bless {
	    	%$config,
	    }, $class;
	}
}

{
	package CactiConfig;

	sub new {
	    my ($class) = @_;

		my $config = Getperf::Config::read_config('getperf_cacti.json');
	    bless {
	    	%$config,
	    }, $class;
	}
}

{
	package ZabbixConfig;

	sub new {
	    my ($class) = @_;

		my $config = Getperf::Config::read_config('getperf_zabbix.json');
	    bless {
	    	%$config,
	    }, $class;
	}
}

{
	package GraphiteConfig;

	sub new {
	    my ($class) = @_;

		my $config = Getperf::Config::read_config('getperf_graphite.json');
	    bless {
	    	%$config,
	    }, $class;
	}
}

{
	package InfluxConfig;

	sub new {
	    my ($class) = @_;

		my $config = Getperf::Config::read_config('getperf_influx.json');
	    bless {
	    	%$config,
	    }, $class;
	}
}

register 'base'     => sub { BaseConfig->new; };
register 'rsync'    => sub { RsyncConfig->new; };
register 'rrd'      => sub { RRDConfig->new; };
register 'queue'    => sub { QueueConfig->new; };
register 'cacti'    => sub { CactiConfig->new; };
register 'zabbix'   => sub { ZabbixConfig->new; };
register 'graphite' => sub { GraphiteConfig->new; };
register 'influx'   => sub { InfluxConfig->new; };
register 'sites'    => sub { SiteConfig->new; };

1;
